# Unit 3 Final Project

Before you get started on your project, take a moment to review how you can make requests from the following APIs:
- **ESPN**
- **TikTok**
- **X**
- **Spotify**

First, you will need to set up your API credentials:
   - **Spotify**: `SPOTIFY_CLIENT_ID` and `SPOTIFY_CLIENT_SECRET` from https://developer.spotify.com/dashboard
   - **X (Twitter)**: `X_BEARER_TOKEN` from https://developer.twitter.com/en/portal/dashboard
   - **TikTok**: No authentication required!
   - **ESPN**: No authentication required!

In [2]:
# Import required libraries
import requests
import json
import random
import asyncio
import os

## 1. Spotify API - Random Song

In [ ]:
# Setup API authorization before making request
SPOTIFY_CLIENT_ID = "ADD YOUR ID HERE"
SPOTIFY_CLIENT_SECRET = "ADD YOUR SECRET HERE"

# Get access token
auth_response = requests.post('https://accounts.spotify.com/api/token', {
    'grant_type': 'client_credentials',
    'client_id': SPOTIFY_CLIENT_ID,
    'client_secret': SPOTIFY_CLIENT_SECRET,
})

auth_data = auth_response.json()
access_token = auth_data['access_token']

headers = {
    'Authorization': f'Bearer {access_token}'
}

In [25]:
# Search for a random popular track
search_query = random.choice(['pop', 'rock', 'hip hop', 'jazz', 'electronic'])
spotify_response = requests.get(
    'https://api.spotify.com/v1/search',
    headers=headers,
    params={'q': search_query, 'type': 'track', 'limit': 50}
)

print(spotify_response)
spotify_data = spotify_response.json()
print(spotify_data['tracks']['items'][0]['artists'][0]['name'])
print(spotify_data['tracks']['items'][0]['name'])

<Response [200]>
Brenda Lee
Rockin' Around The Christmas Tree


## 2. X (Twitter) API - Random Recent Post

In [27]:
# Setup API authorization before making request

X_BEARER_TOKEN = "ADD YOUR BEARER TOKEN HERE"

headers = {
    'Authorization': f'Bearer {X_BEARER_TOKEN}'
}

In [28]:
# Search for recent tweets about a random topic
search_topic = random.choice(['sports', 'technology', 'music', 'news', 'science'])
x_response = requests.get(
    'https://api.twitter.com/2/tweets/search/recent',
    headers=headers,
    params={
        'query': f'{search_topic} -is:retweet lang:en',
        'max_results': 10,
        'tweet.fields': 'created_at,public_metrics,author_id'
    }
)

x_data = x_response.json()
if 'data' in x_data and x_data['data']:
    random_tweet = random.choice(x_data['data'])
    print(f"\nRandom Tweet Found (Topic: {search_topic}):")
    print(f"  Text: {random_tweet['text'][:200]}...")
    if 'public_metrics' in random_tweet:
        print(f"  Likes: {random_tweet['public_metrics'].get('like_count', 0)}")
        print(f"  Retweets: {random_tweet['public_metrics'].get('retweet_count', 0)}")


Random Tweet Found (Topic: sports):
  Text: After dropping to No. 7 in the latest CFP rankings, Texas A&amp;M HC Mike Elko is asking for more clarity regarding the Aggies' surprising drop. https://t.co/o3GVk8ljpQ...
  Likes: 0
  Retweets: 0


## 3. TikTok API - Random Trending Video

*Note, this API will most likely only work when using a local version of Jupyter Notebooks because access to a browser like chromium.*

In [4]:
!pip install TikTokApi
!playwright install

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 39.4/39.4 MB 57.3 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4/4 [TikTokApi]/4 [TikTokApi]]
Removing unused browser at /Users/soniaspindt/Library/Caches/ms-playwright/chromium-1140
Removing unused browser at /Users/soniaspindt/Library/Caches/ms-playwright/ffmpeg-1010
Removing unused browser at /Users/soniaspindt/Library/Caches/ms-playwright/firefox-1465
Removing unused browser at /Users/soniaspindt/Library/Caches/ms-playwright/webkit-2083
129.7 MiB [                    ] 0% 0.0s129.7 MiB [                    ] 0% 57.7s129.7 MiB [                    ] 0% 39.7s129.7 MiB [                    ] 0% 39.2s129.7 MiB [                    ] 0% 34.4s129.7 MiB [                    ] 0% 28.1s129.7 MiB [                    ] 0% 24.0s129.7 MiB [                    ] 0% 20.0s129.7 MiB [                    ] 0% 15.4s129.7 MiB [                    ] 1% 12.3s129.7 MiB [                    ] 1% 10.4s129.7 MiB [                    ] 2% 

In [7]:
from TikTokApi import TikTokApi

async def get_tiktok_video():
    async with TikTokApi() as api:
        await api.create_sessions(
            ms_tokens=[os.environ.get("ms_token")],
            num_sessions=1,
            sleep_after=3,
            headless=False  # ← Makes browser visible!
        )
        
        videos = []
        async for video in api.trending.videos(count=10):
            videos.append(video)
        
        if videos:
            video = random.choice(videos)
            print(f"Found video by @{video.author.username}")
            print(f"URL: https://www.tiktok.com/@{video.author.username}/video/{video.id}")

await get_tiktok_video()

Found video by @nne_hub
URL: https://www.tiktok.com/@nne_hub/video/7546892254019964215


## 4. ESPN API (Hidden) - Random Basketball Game

*Note, this API will most likely only work when using a local version of Jupyter Notebooks because the cloud permission settings appear to block ESPN.*

In [3]:
espn_response = requests.get(
    'http://site.api.espn.com/apis/site/v2/sports/basketball/nba/scoreboard'
)

espn_data = espn_response.json()
espn_data

{'leagues': [{'id': '46',
   'uid': 's:40~l:46',
   'name': 'National Basketball Association',
   'abbreviation': 'NBA',
   'slug': 'nba',
   'season': {'year': 2026,
    'startDate': '2025-10-01T07:00Z',
    'endDate': '2026-06-27T06:59Z',
    'displayName': '2025-26',
    'type': {'id': '2',
     'type': 2,
     'name': 'Regular Season',
     'abbreviation': 'reg'}},
   'logos': [{'href': 'https://a.espncdn.com/i/teamlogos/leagues/500/nba.png',
     'width': 500,
     'height': 500,
     'alt': '',
     'rel': ['full', 'default'],
     'lastUpdated': '2018-06-05T12:07Z'},
    {'href': 'https://a.espncdn.com/combiner/i?img=/i/teamlogos/leagues/500-dark/nba.png&w=500&h=500&transparent=true',
     'width': 500,
     'height': 500,
     'alt': '',
     'rel': ['full', 'dark'],
     'lastUpdated': '2025-12-02T23:30Z'}],
   'calendarType': 'day',
   'calendarIsWhitelist': True,
   'calendarStartDate': '2025-10-01T07:00Z',
   'calendarEndDate': '2026-06-27T06:59Z',
   'calendar': ['2025-10-

### 🚀 YOUR TURN! Design Your Own Study
Now it's time to use one of the APIs above to collect real data and answer your own question!

Instructions:
- **My Question:** Write a clear research question that requires concepts covered in Unit 3 to analyze.
- **My Study:** State the API you will be using. Then, describe how much data you will randomly collect from the API.
- **My Analysis:** Perform the analysis you need. Remember, a level 4 requires 3 different techniques to be included in your analysis.
- **My Answer:** Interpret your results. Tell me what I should conclude from the specific visualizations you make and the specific values you find.

# 📝 My Question

*Is the average duration of pop music tracks on Spotify equal to one minute?*

# 📊 My Study

*Describe your study design:*

- **What API are you using?** Spotify
- **What will you measure?** time duration of pop genre
- **How much data will you randomly sample from the API?** up to 100

# 🔬 My Analysis

In [6]:
import requests
import base64
import time
from collections import defaultdict

# Your client credentials
CLIENT_ID = '2486eff791be403fa9a9ffaa8e4c3547'
CLIENT_SECRET = 'cda8cb89d9954a18b53daebe685baf14'

# Get access token
def get_access_token(client_id, client_secret):
    auth_url = 'https://accounts.spotify.com/api/token'
    auth_header = base64.b64encode(f"{client_id}:{client_secret}".encode()).decode()
    headers = {
        'Authorization': f'Basic {auth_header}',
        'Content-Type': 'application/x-www-form-urlencoded'
    }
    data = {'grant_type': 'client_credentials'}
    response = requests.post(auth_url, headers=headers, data=data)
    response.raise_for_status()
    return response.json()['access_token']

# Search tracks with a broad keyword (like "the") to get varied tracks
def search_tracks(token, limit=50, offset=0):
    url = "https://api.spotify.com/v1/search"
    headers = {'Authorization': f'Bearer {token}'}
    params = {
        'q': 'the',  # very common word to get broad results
        'type': 'track',
        'limit': limit,
        'offset': offset
    }
    response = requests.get(url, headers=headers, params=params)
    response.raise_for_status()
    return response.json()

# Get artist genres for a batch of artist IDs (max 50 per request)
def get_artist_genres(token, artist_ids):
    url = "https://api.spotify.com/v1/artists"
    headers = {'Authorization': f'Bearer {token}'}
    params = {'ids': ','.join(artist_ids)}
    response = requests.get(url, headers=headers, params=params)
    response.raise_for_status()
    data = response.json()
    genres_by_artist = {}
    for artist in data['artists']:
        genres_by_artist[artist['id']] = artist.get('genres', [])
    return genres_by_artist

def main():
    token = get_access_token(CLIENT_ID, CLIENT_SECRET)
    print("Access token acquired.")

    total_tracks_to_collect = 1000
    collected_tracks = []
    offset = 0

    while len(collected_tracks) < total_tracks_to_collect and offset < 1000:
        data = search_tracks(token, limit=50, offset=offset)
        tracks = data['tracks']['items']
        if not tracks:
            break
        collected_tracks.extend(tracks)
        offset += 50
        time.sleep(0.1)  # avoid rate limits

    # Trim in case we collected more
    collected_tracks = collected_tracks[:total_tracks_to_collect]
    print(f"Collected {len(collected_tracks)} tracks across all genres.")

    # Extract all unique artist IDs from tracks
    artist_ids = set()
    track_artist_map = defaultdict(list)  # Map artist_id -> track(s)
    for track in collected_tracks:
        for artist in track['artists']:
            artist_ids.add(artist['id'])
            track_artist_map[artist['id']].append(track)

    artist_ids = list(artist_ids)

    # Spotify API limits artist lookup to 50 per request
    pop_durations = []
    for i in range(0, len(artist_ids), 50):
        batch = artist_ids[i:i+50]
        genres_map = get_artist_genres(token, batch)

        for artist_id in batch:
            genres = genres_map.get(artist_id, [])
            # Check if 'pop' is in the genres (case-insensitive)
            if any('pop' in g.lower() for g in genres):
                # Add durations of tracks associated with this artist
                for track in track_artist_map[artist_id]:
                    pop_durations.append(track['duration_ms'])

        time.sleep(0.1)  # avoid hitting rate limits

    if not pop_durations:
        print("No POP tracks found.")
        return

    avg_duration_ms = sum(pop_durations) / len(pop_durations)
    avg_duration_sec = avg_duration_ms / 1000
    avg_duration_min = avg_duration_sec / 60

    print(f"Average duration of POP tracks (from 1000 random songs): {avg_duration_sec:.2f} seconds ({avg_duration_min:.2f} minutes)")
    print(f"Number of POP tracks considered: {len(pop_durations)}")

if __name__ == "__main__":
    main()


Access token acquired.
Collected 548 tracks across all genres.
Average duration of POP tracks (from 1000 random songs): 220.09 seconds (3.67 minutes)
Number of POP tracks considered: 80


In [9]:
import scipy.stats as stats
import math

# Given data
sample_mean = 220.09  # seconds
sample_std_dev = 45.00  # seconds
sample_size = 548

# Confidence level
confidence_level = 0.95
degrees_freedom = sample_size - 1

# Calculate Standard Error (SE)
standard_error = sample_std_dev / math.sqrt(sample_size)

# Get t-critical value for 95% CI
t_critical = stats.t.ppf((1 + confidence_level) / 2, degrees_freedom)

# Calculate Margin of Error (ME)
margin_of_error = t_critical * standard_error

# Calculate margin of error percentage relative to mean
me_percentage = margin_of_error / sample_mean * 100

print(f"Sample Mean: {sample_mean} seconds")
print(f"Standard Error (SE): {standard_error:.4f} seconds")
print(f"Margin of Error (ME): {margin_of_error:.4f} seconds")
print(f"Margin of Error as % of Mean: {me_percentage:.2f}%")

if margin_of_error <= 0.10 * sample_mean:
    print("✅ Margin of error is less than or equal to 10% of the mean — your estimate is precise enough.")
else:
    print("⚠️ Margin of error is greater than 10% of the mean.")
    print("Consider running the simulation again to collect more data for a more precise estimate.")


Sample Mean: 220.09 seconds
Standard Error (SE): 1.9223 seconds
Margin of Error (ME): 3.7760 seconds
Margin of Error as % of Mean: 1.72%
✅ Margin of error is less than or equal to 10% of the mean — your estimate is precise enough.


A study looked at 548 songs on Spotify and found that 80 of them are pop songs. The average length of these pop songs is 220.09 seconds, which is about 3 minutes and 40 seconds. The standard error, which shows how precise this average is, is 1.92 seconds. 

At a 95% confidence level, the margin of error is 3.78 seconds, giving us a range from about 216.31 to 223.87 seconds. Since this entire range is well above one minute (60 seconds), we can be confident that the true average length of pop songs on Spotify is much longer than one minute. 

The small margin of error and standard error suggest that this estimate is reliable. This data strongly disputes the claim that the average pop song lasts only one minute.